plutus-tx/src/PlutusTx/Data/List.hs

```diff
@@ -212,4 +256,133 @@ mconcat (List l) = go l
             )
 {-# INLINABLE mconcat #-}
 
+uncons :: (UnsafeFromData a) => List a -> Maybe (a, List a)
+uncons (List l) = do
+    (h, t) <- B.uncons l
+    pure (unsafeFromBuiltinData h, List t)
+{-# INLINEABLE uncons #-}
+
+and :: List Bool -> Bool
+and (List l) = go l
+  where
+    go =
+        B.caseList'
+            True
+            (\h t ->
+                let h' = unsafeFromBuiltinData h
+                in h' && go t
+            )
+{-# INLINEABLE and #-}
+
+or :: List Bool -> Bool
+or (List l) = go l
+  where
+    go =
+        B.caseList'
+            False
+            (\h t ->
+                let h' = unsafeFromBuiltinData h
+                in h' || go t
+            )
+{-# INLINEABLE or #-}
+
+elem :: (ToData a) => a -> List a -> Bool
+elem x (List l) = go l
+  where
+    go =
+        B.caseList'
+            False
+            (\h t ->
+                toBuiltinData x == h || go t
+            )
+{-# INLINEABLE elem #-}
+
+notElem :: (ToData a) => a -> List a -> Bool
+notElem x (List l) = not $ elem x (List l)
+{-# INLINEABLE notElem #-}
+
+foldr :: (UnsafeFromData a) => (a -> b -> b) -> b -> List a -> b
+foldr f z (List l) = go l z
+  where
+    go =
+        B.caseList'
+            id
+            (\h t ->
+                let h' = unsafeFromBuiltinData h
+                in f h' . go t
+            )
+{-# INLINEABLE foldr #-}
+
+foldl :: (UnsafeFromData a) => (b -> a -> b) -> b -> List a -> b
+foldl f z (List l) = go z l
```

Comment:
It's unnecessary eta-expansion, you could've used something like `foldl f z = coerce (go z)`. May require some type annotations.

Reply:
Good point, should be fixed (almost) everywhere.